Big Bang Component Extension (#1094)
defenseunicorns/zarf: adr/0012-bigbang-as-a-noun.md

# 12. BigBang as a Noun

Date: 2023-01-18

## Status

Accepted

## Context

> One primary application component that end users of Zarf are deploying is [Big Bang](https://repo1.dso.mil/platform-one/big-bang/bigbang).  The installation of BigBang is complicated for several reason:

Reviewer: ```suggestion
One primary application component that end users of Zarf are deploying is [Big Bang](https://repo1.dso.mil/big-bang/bigbang).  The installation of BigBang is complicated for several reason:
```

> - It requires Flux to be installed to deploy correctly do to the use of Flux CRDs. 

Reviewer 2: ```suggestion
- It requires Flux to be installed to deploy correctly due to the use of Flux CRDs. 
```

and the discovery of which images are needed is a function of the values that are provided to the BigBang chart

## Decision

Deployments of BigBang can be managed with a new `bigbang` noun in the zarf.yaml that manages the complexity of the deployment.  This capability will take the values provided to the big bang chart, template them during the package phase to identify which [BigBang packages](https://repo1.dso.mil/big-bang/bigbang/-/blob/master/docs/packages.md) are being configured in the Zarf package.  The code then includes only the git repositories and images needed for the configured packages, and does not include the git repositories and images for packages that would not be deployed.  

 The `bigbang` section will provide the following configurations for managing a big bang deployment:

- `version` - Identifies the particular version of Bigbang to deploy, which correspond to git tags in the provided `repo`.  See versions of BigBang [here](https://repo1.dso.mil/big-bang/bigbang/-/releases).  
- `repo` - Identifies the git repository BigBang is hosted on.  Defaults to https://repo1.dso.mil/platform-one/big-bang/bigbang.git

> - `valuesFrom` - list of local files that get passed to the BigBang helm chart for deployment. 

Reviewer 2: ```suggestion
- `valuesFrom` - list of local files that get added to the zarf package at create time which will be applied to the Big Bang helm chart. Deploy time variables are able to be declared as part of said values files.
```

@runyontr does this track with your thinking as well?

Author: Yes, we should have deploy time variables able to be injected into those values.yaml files.  We'd need to make sure they get a value in package so we can pull the correct images/repos though.  @corang is that how its implemented currently?

Reviewer 3: any further comment on this?

- `skipFlux` - boolean to determine if the flux installation for BigBang should be skipped.  Only set this to true if flux has been deployed in a different way already in the cluster.

## Consequences

- By doing package time rendering and discovery of images for inclusion into the zarf package, the flexibility for deploy time configuration is limited since new parts of BigBang can't be added arbitrarily, since the necessary artifacts to deploy those BigBang packages won't be present in the zarf package

> - BigBang is every changing and improving, and while it is available as open source, we do not control the change in how the deployment is handled as [BigBang 2.0](https://repo1.dso.mil/groups/big-bang/-/epics/217) is progressing.  This creates a burdon on the Zarf team to ensure new changes in Big Bang do not break how BigBang is deployed, and a burdon to ensure as the way BigBang gets deployed is changed, it does not break older versions of deploying BigBang.

Reviewer 2: ```suggestion
- BigBang is every changing and improving, and while it is available as open source, we do not control the change in how the deployment is handled as [BigBang 2.0](https://repo1.dso.mil/groups/big-bang/-/epics/217) is progressing.  This creates a burden on the Zarf team to ensure new changes in Big Bang do not break how BigBang is deployed, and a burden to ensure as the way BigBang gets deployed is changed, it does not break older versions of deploying BigBang.
```

> - BigBang is every changing and improving, and while it is available as open source, we do not control the change in how the deployment is handled as [BigBang 2.0](https://repo1.dso.mil/groups/big-bang/-/epics/217) is progressing.  This creates a burden on the Zarf team to ensure new changes in Big Bang do not break how BigBang is deployed, and a burden to ensure as the way BigBang gets deployed is changed, it does not break older versions of deploying BigBang.

Reviewer 4: ```suggestion
- Big Bang is ever changing and improving, and while it is available as open source, we do not control the change in how the deployment is handled as [Big Bang 2.0](https://repo1.dso.mil/groups/big-bang/-/epics/217) is progressing.  This creates a burden on the Zarf team to ensure new changes in Big Bang do not break how Big Bang is deployed, and a burden to ensure as the way BigBang gets deployed is changed, it does not break older versions of deploying Big Bang.
```

> - BigBang is every changing and improving, and while it is available as open source, we do not control the change in how the deployment is handled as [BigBang 2.0](https://repo1.dso.mil/groups/big-bang/-/epics/217) is progressing.  This creates a burden on the Zarf team to ensure new changes in Big Bang do not break how BigBang is deployed, and a burden to ensure as the way BigBang gets deployed is changed, it does not break older versions of deploying BigBang.

Reviewer: ```suggestion
- BigBang is ever changing and improving, and while it is available as open source, we do not control the change in how the deployment is handled as [BigBang 2.0](https://repo1.dso.mil/groups/big-bang/-/epics/217) is progressing.  This creates a burden on the Zarf team to ensure new changes in Big Bang do not break how BigBang is deployed, and a burden to ensure as the way BigBang gets deployed is changed, it does not break older versions of deploying BigBang.
```

Minor tweak and curious - do we want to re-word this to be accurate in the future? IE we reference an in progress epic that will change state. But maybe that language lends itself to the decision making process?

 The `bigbang` section will provide the following configurations for managing a big bang deployment:

- `version` - Identifies the particular version of Bigbang to deploy, which correspond to git tags in the provided `repo`.  See versions of BigBang [here](https://repo1.dso.mil/big-bang/bigbang/-/releases).  
- `repo` - Identifies the git repository BigBang is hosted on.  Defaults to https://repo1.dso.mil/big-bang/bigbang.git
- `valuesFrom` - list of local files that get passed to the BigBang helm chart for deployment. 
- `skipFlux` - boolean to determine if the flux installation for BigBang should be skipped.  Only set this to true if flux has been deployed in a different way already in the cluster.

## Implementation

> The BigBang component in the origional zarf.yaml will be split into up to two different zarf components for packaging and deployment.  The first is the flux component, which will only be included if `skipFlux` is set to false (the default). The flux deployment is just a remote Kustomizatoin pointing at the [suggested installation of flux](https://repo1.dso.mil/big-bang/bigbang/-/tree/master/base/flux):   The BigBang component will be the following objects deployed as manifests:

Reviewer 3: ```suggestion
The BigBang component in the original zarf.yaml will be split into one or two different Zarf components.  The first component, flux, by default will be included unless the `skipFlux` flag is set to true. The flux deployment is just a remote Kustomization pointing at the [corresponding version of flux](https://repo1.dso.mil/big-bang/bigbang/-/tree/master/base/flux):   The BigBang component will be the following objects deployed as manifests:
```

1. The `bigbang` `Namespace`
2. One `Secret` for each file provided in `valuesFrom`

> 3. `GitRepository` object pointing at the provided `repo`.

Reviewer 3: I was going through this PR and stumbled on this point. Decided to check out OCI support for the helm sdk and it seems to be a thing now: https://github.com/defenseunicorns/zarf/pull/1346/files. This won't directly help here because we need to keep it as an OCI thing, but something we should keep in mind: what do we see as the future here for BB, dropping the git repo completely or something else?

https://github.com/defenseunicorns/zarf/pull/1346/files

Author: The vision for BigBang is to move all of those `GitRepository` objects to OCI objects, but I think the plan should be to remain with GitRepository objects until:

1. BigBang supports that format by default
2. The Flux controllers can pull http based OCI objects from the Zarf Registry:  See the blocked issue here: https://github.com/defenseunicorns/zarf/issues/1092

4. `HelmRelease` object deploying the chart from the `GitRepository` configured with the secrets as values files.

## Consequences

> 

Reviewer 3: Probably should run formatting on this document as there are a few line break inconsistencies that don't match prettier I believe.

Author: Do you have a tool for this?
